src/storage/storage-types.ts

```diff
@@ -26,14 +22,63 @@ import {
 } from '../util/errors';
 
 //================================================================================
+// TYPES AND EVENTS
 
 export type StorageId = string;
 
-export type StorageEvent =
-    'ingest' |  // 'ingest|/some/path.txt'
-    'willClose' | 'didClose';
+export type StorageBusChannel =
+    'ingest' |  // 'write|/some/path.txt'  // note that write errors and no-ops are also sent here
+    'willClose' |
+    'didClose';
+
+export interface QueryResult {
+    // the docs from the query...
+    docs: Doc[],
+    // ...and the storageDriver's maxLocalIndex at the time the
+    // query was done.  This is the OVERALL max local index for
+    // the whole storage, not just for the resulting docs.
+    // TODO: include the max for the doc results also
+    maxLocalIndex: number,
+}
+
+// IngestEvents are returned from storage.set() and storage.ingest(),
+// and sent as events on the storage.bus 'ingest' channel.
+
+export interface IngestEventFailure {
+    kind: 'failure',
+    reason: 'write_error' | 'invalid_document',
+    maxLocalIndex: number,
+    err: Error | null,
+}
+export interface IngestEventNothingHappened {
+    kind: 'nothing_happened',
+    reason: 'obsolete_from_same_author' | 'already_had_it'
+    maxLocalIndex: number,
+    doc: Doc,  // won't have a _localIndex because it was not actually ingested
+}
+export interface IngestEventSuccess {
+    kind: 'success',
+    maxLocalIndex: number,
+    doc: Doc,  // the just-written doc, frozen, with updated extra properties like _localIndex
+
+    docIsLatest: boolean,  // is it the latest at this path (for any author)?
+
+    // the most recent doc from the same author, at this path, before the new doc was written.
+    prevDocFromSameAuthor: Doc | null,
+
+    // the latest doc from any author at this path, before the new doc was written.
+    // note this is actually still the latest doc if the just-written doc is an older one (docIsLatest===false)
+    prevLatestDoc: Doc | null,
+}
+export type IngestEvent =
+    IngestEventFailure |
+    IngestEventNothingHappened |
+    IngestEventSuccess;
```

Comment:
Here's the new `IngestEvent` that is returned from `storage.ingest(doc)`, and also sent out on the storage bus.